templates/default/haproxy.cfg.erb

```diff
@@ -1,40 +1,149 @@
-  stats socket <%= node['haproxy']['stats_socket_path'] -%> <%= node['haproxy']['stats_socket'].map{|k, v| "#{k} #{v}"}.flatten.join(' ') %>
-<% end -%>
-<% node['haproxy']['global_options'].sort.each do |option, value| %>
+  user <%= @global['user'] %>
+  group <%= @global['group'] %>
+  <% if @global['pid_file'] -%>pidfile <% @global['pid_file'] %><% end %>
+  log <%= @global['log'] %>
+  <% if @global['log_tag'] -%>log-tag <%= @global['log_tag'] %><% end -%>
+  <% if @global['chroot'] -%>chroot <%= @global['chroot'] %><% end -%>
+  <% if @global['daemon'].eql? 'true' -%>daemon<% end -%>
+  <%= @global['debug_option'] %>
+  stats socket <%= @global['stats_socket'] %>
+  stats timeout <%= @global['stats_timeout'] %>
+  <% if @global['pidfile'] -%>pidfile <%= @global['pidfile'] %><% end -%>
+  <% unless @global['tuning'].nil? %>
+  <% @global['tuning'].each do | option, value | -%>
+  tune.<%= option %> <%= value %>
+  <% end -%>
+  <% end -%>
+  <% unless @global['extra_options'].nil? -%>
+  <% @global['extra_options'].each do | option, value | -%>
   <%= option %> <%= value %>
-<% end %>
+  <% end -%>
+  <% end -%>
+<% end -%>
 
+<% unless @defaults.nil? %>
 defaults
-  log     global
-  mode    <%= node['haproxy']['mode'] %>
-  retries <%= node['haproxy']['defaults_retries'] %>
-<% @defaults_timeouts.sort.map do | value, time | -%>
-  timeout <%= value %> <%= time %>
-<% end -%>
-<% @defaults_options.sort.each do | option | -%>
+  <% @defaults['timeout'].each do | key, timeout | -%>
+  timeout <%= key %> <%= timeout %>
+  <% end -%>
+  log <%= @defaults['log'] %>
+  mode <%= @defaults['mode'] %>
+  <% unless @defaults['balance'].empty? -%>balance <%= @defaults['balance'] %><% end -%>
+  <% @defaults['option'].each do | option | -%>
   option <%= option %>
+  <% end -%>
+
+  <% unless @defaults['maxconn'].nil? -%>maxconn <%= @defaults['maxconn'] %><% end -%>
+  <% unless @defaults['retries'].nil? -%>retries <%= @defaults['retries'] %><% end -%>
+
+  <% unless @defaults['stats'].nil? %>
+  <% @defaults['stats'].each do | key, value | -%>
+  stats <%= key %> <%= value %>
+  <% end -%>
+  <% end -%>
+
+  <% unless @defaults['extra_options'].nil? %>
+  <% @defaults['extra_options'].each do | option, value | -%>
+  <%= option %> <%= value %>
+  <% end -%>
+  <% end -%>
 <% end -%>
-  balance  <%= node['haproxy']['balance_algorithm'] %>
 
-# Set up application listeners here.
+<% unless @frontend.nil? %>
+<% @frontend.each do |frontend, f | %>
+frontend <%= frontend %>
+  default_backend <%= f['default_backend'] %>
+  <% unless f['bind'].empty? -%>bind <%= f['bind'] %><% end -%>
+  <% unless f['maxconn'].empty? -%>maxconn <%= f['maxconn'] %><% end -%>
+  <% unless f['http_request'].empty? -%>http_request<%= f['http_request'] %><% end -%>
+  <% unless f['http_response'].empty? -%>http_response<%= f['http_response'] %><% end -%>
+
+  <% unless f['use_backend'].nil? %>
+  <% f['use_backend'].each do | backend |%>
+  <% backend.each do | backend | %>
```

Comment:
Having the double each here seems odd to me, should this be `f['use_backend'].flatten.each do |backend|`